ocaml/typing/ctype.ml

```diff
@@ -2195,137 +2125,107 @@ let type_jkind_sub env ty jkind =
-           infinitely expanding a recursive type. In a wide product, we many
-           need to expand many types shallowly, and that's fine. *)
-        let results = List.map2 (check_one fuel) component_types jkinds in
-        process_product_results fuel [] [] results
-
-  (* Here, we are recusing down [results] is a list of the (unprocessed) results
-     from attempting to constrain a product jkind. There is one for each
-     component, and we process them individually into the accumulators [vars]
-     and [jkind_ests]. Those collect the [Type_vars] from the component checks
-     and the jkind estimate of each component, respectively
-
-     [Failure]s in results may have occurred just because we haven't expanded
-     the component type enough yet, so we expand and check it again. *)
-
-  and process_product_results fuel vars jkind_ests results =
-    match results with
-    | [] ->
-      let jkind =
-        Jkind.Builtin.product ~why:Unboxed_tuple (List.rev jkind_ests)
-      in
-      if vars = [] then jkind, Success else jkind, Type_vars vars
-    | (jkind_est, result) :: results ->
-      (* If the result is failure, we try again with expansion. *)
-      let jkind_est, result =
-        match result with
-        (* CR: We could refactor slightly so that we skip the shortcut check
-           in [expand_and_check_one] here. *)
-        | Failure { bound; ty } -> expand_and_check_one fuel ty bound
-        | Success | Type_vars _ | Missing_cmi _ -> jkind_est, result
-      in
-      let jkind_ests = jkind_est :: jkind_ests in
-      match result with
-      | Success -> process_product_results fuel vars jkind_ests results
-      | Type_vars vars' ->
-        process_product_results fuel (vars' @ vars) jkind_ests results
-      | Failure _ | Missing_cmi _ ->
-        let jkind = jkind_est_of_results jkind_ests results in
-        jkind, result
-  in
-  expand_and_check_one 100 ty jkind
+    | Tvar _ when not fixed ->
+       (* Unfixed tyvars are special in at least two ways:
+
+          1) Suppose we're processing [type 'a t = 'a list]. The ['a] on the
+          left will be born with an [Unannotated_type_parameter] history and a
+          sort variable as its jkind. When unifying with the arguments to
+          [list], we'll need to change that jkind to be [value]. But we want to
+          do this update by changing the type variable, not just by filling in
+          the sort variable: the former preserves history (improving error
+          messages), while the latter does not.  (If you're readying this in a
+          future where sorts track their own history, then this reason no longer
+          applies; delete this bullet.)
+
+          2) Suppose we're unifying ['a : value] with ['a : float64]. This would
+          be possible if those two jkinds had an intersection. And so an error
+          message should complain about the lack of intersection, not the lack
+          of subjkinding.
```

Comment:
Ah -- this point isn't new. That is, we were already (correctly) complaining about intersections. My patch now doesn't do this as aggressively as we did previously (because it does less expansion). Without this special case, the regression (complaining about subkinding instead of intersections) would be worse.

I've rephrased not to be about unification.